pkg/endpoint/policy.go

```diff
@@ -227,123 +175,146 @@ func (e *Endpoint) applyL4PolicyLocked(owner Owner, oldPolicy *policy.L4Policy,
-	//if c.Iteration == cache.Iteration {
-	//	repo.Mutex.RUnlock()
-	//	e.getLogger().WithField("consumableID", c.ID).Debug("Reusing cached policy for consumable identity")
-	//	return false, nil
-	//}
+	reservedIDs := policy.GetConsumableCache().GetReservedIDs()
+	var idx policy.NumericIdentity
+	for _, idx = range reservedIDs {
+		lbls, err := owner.GetCachedLabelList(idx)
+		if err != nil {
+			return nil, err
+		}
+		// Skip currently unused IDs
+		if lbls == nil || len(lbls) == 0 {
+			continue
+		}
+		labelsMap[idx] = lbls
+	}
 
-	maxID, err := owner.GetCachedMaxLabelID()
-	if err != nil {
-		return false, err
+	for idx = policy.MinimalNumericIdentity; idx < maxID; idx++ {
+		lbls, err := owner.GetCachedLabelList(idx)
+		if err != nil {
+			return nil, err
+		}
+		// Skip currently unused IDs
+		if lbls == nil || len(lbls) == 0 {
+			continue
+		}
+		labelsMap[idx] = lbls
 	}
 
-	c.Mutex.RLock()
+	return &labelsMap, nil
+}
+
+// Must be called with global endpoint.Mutex held
+func (e *Endpoint) resolveL4Policy(owner Owner, repo *policy.Repository, c *policy.Consumable) error {
 	ctx := policy.SearchContext{
 		To: c.LabelArray,
 	}
-	c.Mutex.RUnlock()
-
 	if owner.TracingEnabled() {
 		ctx.Trace = policy.TRACE_ENABLED
 	}
 
-	repo := owner.GetPolicyRepository()
-	repo.Mutex.Lock()
-	newL4policy, err := repo.ResolveL4Policy(&ctx)
-	defer repo.Mutex.Unlock()
-
+	newL4Policy, err := repo.ResolveL4Policy(&ctx)
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	c.Mutex.Lock()
-	defer c.Mutex.Unlock()
+	if reflect.DeepEqual(c.L4Policy, newL4Policy) {
+		return nil
+	}
+
+	c.L4Policy = newL4Policy
+	return nil
+}
+
+// Must be called with global endpoint.Mutex held
+func (e *Endpoint) regenerateConsumable(owner Owner, labelsMap *LabelsMap, repo *policy.Repository, c *policy.Consumable) bool {
+	changed := false
+
 	// Mark all entries unused by denying them
 	for k := range c.Consumers {
 		c.Consumers[k].DeletionMark = true
 	}
 
-	if c.L4Policy != nil {
-		e.cleanUnusedRedirects(owner, c.L4Policy.Ingress, newL4policy.Ingress)
-		e.cleanUnusedRedirects(owner, c.L4Policy.Egress, newL4policy.Egress)
-	}
+	// L4 policy needs to be applied on two conditions
+	// 1. The L4 policy has changed
+	// 2. The set of applicable security identities has changed.
+	if e.L4Policy != c.L4Policy || e.LabelsMap != labelsMap {
+		// PolicyMap can't be created in dry mode.
+		if !owner.DryModeEnabled() {
+			// Update Endpoint's L4Policy
+			if e.L4Policy != nil {
+				e.cleanUnusedRedirects(owner, e.L4Policy.Ingress, c.L4Policy.Ingress)
+				e.cleanUnusedRedirects(owner, e.L4Policy.Egress, c.L4Policy.Egress)
+			}
 
-	err = e.applyL4PolicyLocked(owner, c.L4Policy, newL4policy)
-	if err != nil {
-		return false, err
+			err := e.applyL4PolicyLocked(labelsMap, e.L4Policy, c.L4Policy)
+			if err != nil {
+				// This should not happen, and we can't fail at this stage anyway.
+				e.getLogger().Fatal("L4 Policy application failed")
```

Comment:
Do we want to run `Fatal` here? This will crash the daemon?

Reply:
This can fail only if the bpf map update fails, and that is an error we can not recover from, and it may be better to have cilium restart.

That said, I hope that in future we can push bpf map updates one layer down for policy computation.

Reply:
Thanks for clarifying, makes sense.